### PR TITLE
Added forward handler

### LINKED ISSUE #696: Cant catch forwarded messages
If i forward message to bot, I cant catch it. Message text is "" and filled fields are Origin, Caption and some others.

### PR DESCRIPTION
**Pull request by nikepan**
Resolve #696

### CODE REVIEW

File: update.go [hunk 1]
```diff
@@ -74,6 +74,10 @@ func (b *Bot) ProcessUpdate(u Update) {
 			return
 		}
 
+		if m.Origin != nil && m.Text == "" {
```
Comment (demget) [inline update.go]: Why is the text should be empty?

**Review by demget — CHANGES_REQUESTED** [review]

File: update.go (hunk 1 above)
Comment (nikepan) [reply update.go]: It seems to be always empty when forwarding. but if it is for example a reply, the text will not be empty.

Comment (demget) [reply update.go]: The emptiness of `m.Text` is already checked here

https://github.com/tucnak/telebot/blob/350ac26ac84216662c19baea128c18fecbc0eb84/update.go#L43

Comment (demget) [reply update.go]: Could you please also add the same event for reply to a message? There is a `ReplyTo` field in the `Message`

**Review by demget — CHANGES_REQUESTED** [review]

**Comment by nikepan** [comment]
Here ok? https://github.com/tucnak/telebot/blob/350ac26ac84216662c19baea128c18fecbc0eb84/update.go#L68

**Comment by demget** [comment]
> Here ok?
> 
> https://github.com/tucnak/telebot/blob/350ac26ac84216662c19baea128c18fecbc0eb84/update.go#L68

Yes

**Comment by nikepan** [comment]
I made some tests again. text="" if it forward from channel. if it from chat, text is original and trigger text events.
I tested reply event. It works fine.

**Comment by demget** [comment]
> I made some tests again. text="" if it forward from channel. if it from chat, text is original and trigger text events. I tested reply event. It works fine.

Then you should move the `m.Origin` check before the `m.Text` check. Complete this and I'll merge your PR into `v3.4`.

**Comment by nikepan** [comment]
I think this action will create a backwards incompatibility for applications that recognise forwarding from chats now as a normal message

**Comment by demget** [comment]
> I think this action will create a backwards incompatibility for applications that recognise forwarding from chats now as a normal message

Oh yeah, you're definitely right. We should check somehow if there is `OnForward` or `OnReply` handler registered, before trying to execute them, if not, the bot should follow the usual `OnText` flow.

Do you think you can implement it in a separate PR?

**Comment by nikepan** [comment]
I don't think that's a good idea. Because it violates the text filters.  I took another look. The current version doesn't break anything if there is no OnForward handler. But if there is one, 2 events are triggered. I think the most correct thing would be to move the check down. Then it will be possible to catch the event if text filters are not triggered, which I believe should be a priority.

**Pull request merged**